pyduofern/duofern.py

```diff
@@ -1007,46 +1017,46 @@ def set(self, code, cmd, *args):
             if subCmd not in commands[cmd]:
                 raise Exception("Wrong argument {}, {}".format(arg, subCmd))
 
-            position = -1 if not "position" in self.modules['by_code'][code] else self.modules['by_code'][code][
+            position = -1 if not "position" in self.modules['by_code'][device_id] else self.modules['by_code'][device_id][
                 "position"]
             # toggleUpDown    = AttrVal(name, "toggleUpDown", "0")
-            toggleUpDown = self.modules['by_code'][code]['toggleUpDown'] if 'toggleUpDown' in self.modules['by_code'][
-                code] else 0
-            moving = "stop" if not "moving" in self.modules['by_code'][code] else self.modules['by_code'][code][
+            toggleUpDown = self.modules['by_code'][device_id]['toggleUpDown'] if 'toggleUpDown' in self.modules['by_code'][
+                device_id] else 0
+            moving = "stop" if not "moving" in self.modules['by_code'][device_id] else self.modules['by_code'][device_id][
                 "moving"]
-            timeAutomatic = "on" if not "timeAutomatic" in self.modules['by_code'][code] else \
-                self.modules['by_code'][code]["timeAutomatic"]
-            dawnAutomatic = "on" if not "dawnAutomatic" in self.modules['by_code'][code] else \
-                self.modules['by_code'][code]["dawnAutomatic"]
-            duskAutomatic = "on" if not "duskAutomatic" in self.modules['by_code'][code] else \
-                self.modules['by_code'][code]["duskAutomatic"]
+            timeAutomatic = "on" if not "timeAutomatic" in self.modules['by_code'][device_id] else \
+                self.modules['by_code'][device_id]["timeAutomatic"]
+            dawnAutomatic = "on" if not "dawnAutomatic" in self.modules['by_code'][device_id] else \
+                self.modules['by_code'][device_id]["dawnAutomatic"]
+            duskAutomatic = "on" if not "duskAutomatic" in self.modules['by_code'][device_id] else \
+                self.modules['by_code'][device_id]["duskAutomatic"]
 
             if moving != "stop":
                 if cmd in ('up', 'down', 'toggle'):
                     if toggleUpDown:
                         cmd = "stop"
-            # self.update_state(code,"moving","moving")
+            # self.update_state(hash,"moving","moving")
 
             if ((cmd == "toggle") and (position > -1)):
-                self.update_state(code, "moving", "moving", 1)
+                self.update_state(hash, "moving", "moving", 1)
```

Comment:
The travis timeout happens because in this line ``hash`` is undefined (same issue with all other references to ``hash`` in this method). **To merge upstream we have to fix this.**

Unfortunately just going back to using ``code`` or ``device_id`` here does not help because ``update_state`` was refactored to expect the module_definition instead of a device id. It may make sense to leave this as is and instead revert to the previous behaviour of ``update_state`` and instead adapting the ``code``.